File: src/main/java/com/tkr/desk/AllocationSolver.java
package com.tkr.desk;

import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/** Min/max constraint solver adjusting pro-rata slices via deficit redistribution. */
public final class AllocationSolver {

    public static final class AllocationBound {
        public int accountId;
        public int minQtyMilli;
        public int maxQtyMilli;
    }

    public static final class AllocationSolverConfig {
        public boolean iterativeRedistribute = true;
        public int maxIterations = 100;
    }

    public static final class AllocationSolverResult {
        public Status status = Status.OK;
        public List<ProRataAllocator.ProRataSlice> finalSlices = new ArrayList<>();
        public boolean converged;
        public int iterations;
    }

    private final AllocationSolverConfig config;
    private final List<AllocationBound> customBounds = new ArrayList<>();

    public AllocationSolver() { this(new AllocationSolverConfig()); }
    public AllocationSolver(AllocationSolverConfig config) {
        this.config = config != null ? config : new AllocationSolverConfig();
    }

    public void setBound(int accountId, int minQty, int maxQty) {
        for (AllocationBound b : customBounds) {
            if (b.accountId == accountId) { b.minQtyMilli = minQty; b.maxQtyMilli = maxQty; return; }
        }
        AllocationBound bound = new AllocationBound();
        bound.accountId = accountId;
        bound.minQtyMilli = minQty;
        bound.maxQtyMilli = maxQty;
        customBounds.add(bound);
    }

    public void clearBounds() { customBounds.clear(); }

    public AllocationSolverResult solve(com.tkr.types.WireTypes.BatchWireFrame frame,
                                        List<ProRataAllocator.ProRataSlice> slices) {
        int targetTotal = 0;
        if (frame != null) {
            for (com.tkr.types.WireTypes.WireBatchRecord rec : frame.records) targetTotal += rec.qtyMilli;
        }
        return solveWithBounds(slices, null, targetTotal);
    }

    public AllocationSolverResult solveWithBounds(List<ProRataAllocator.ProRataSlice> slices,
                                                  List<AllocationBound> bounds, int targetTotal) {
        AllocationSolverResult result = new AllocationSolverResult();
        List<AllocationBound> allBounds = new ArrayList<>(customBounds);
        if (bounds != null) allBounds.addAll(bounds);
        List<ProRataAllocator.ProRataSlice> working = deepCopy(slices);
        result.iterations = 0;
        boolean stable = false;
        while (!stable && result.iterations < config.maxIterations) {
            stable = true;
            result.iterations++;
            List<Violation> violations = detectViolations(working, allBounds);
            if (violations.isEmpty()) break;
            for (Violation v : violations) {
                if (v.surplus > 0) {
                    clampSlice(findSlice(working, v.accountId), v.maxQty);
                    Status st = redistributeSurplus(working, v.surplus, v.accountId);
                    if (st != Status.OK) { result.status = st; return result; }
                    stable = false;
                } else if (v.deficit > 0) {
                    clampSlice(findSlice(working, v.accountId), v.minQty);
                    Status st = redistributeDeficit(working, v.deficit, v.accountId);
                    if (st != Status.OK) { result.status = st; return result; }
                    stable = false;
                }
            }
        }
        result.converged = stable || detectViolations(working, allBounds).isEmpty();
        result.finalSlices = working;
        int sum = sumSliceQty(working);
        if (sum != targetTotal) result.status = Status.BOUNDS_ERROR;
        return result;
    }

    private static final class Violation {
        int accountId, currentQty, minQty, maxQty, deficit, surplus;
    }

    private List<Violation> detectViolations(List<ProRataAllocator.ProRataSlice> slices, List<AllocationBound> bounds) {
        List<Violation> out = new ArrayList<>();
        for (ProRataAllocator.ProRataSlice slice : slices) {
            AllocationBound bound = findBound(bounds, slice.accountId);
            if (bound == null) continue;
            Violation v = new Violation();
            v.accountId = slice.accountId;
            v.currentQty = slice.qtyMilli;
            v.minQty = bound.minQtyMilli;
            v.maxQty = bound.maxQtyMilli;
            if (slice.qtyMilli < bound.minQtyMilli) v.deficit = bound.minQtyMilli - slice.qtyMilli;
            if (slice.qtyMilli > bound.maxQtyMilli) v.surplus = slice.qtyMilli - bound.maxQtyMilli;
            if (v.deficit > 0 || v.surplus > 0) out.add(v);
        }
        return out;
    }

    private AllocationBound findBound(List<AllocationBound> bounds, int accountId) {
        for (AllocationBound b : bounds) if (b.accountId == accountId) return b;
        return null;
    }

    private ProRataAllocator.ProRataSlice findSlice(List<ProRataAllocator.ProRataSlice> slices, int accountId) {
        for (ProRataAllocator.ProRataSlice s : slices) if (s.accountId == accountId) return s;
        return null;
    }

    private void clampSlice(ProRataAllocator.ProRataSlice slice, int limit) {
        if (slice == null) return;
        slice.qtyMilli = limit;
    }

    private Status redistributeDeficit(List<ProRataAllocator.ProRataSlice> slices, int deficit, int excludeAccount) {
        if (deficit <= 0) return Status.OK;
        slices.sort(Comparator.comparingInt(s -> s.qtyMilli));
        int remaining = deficit;
        for (int i = slices.size() - 1; i >= 0 && remaining > 0; i--) {
            ProRataAllocator.ProRataSlice s = slices.get(i);
            if (s.accountId == excludeAccount || s.qtyMilli <= 0) continue;
            s.qtyMilli -= 1;
            findSlice(slices, excludeAccount).qtyMilli += 1;
            remaining--;
        }
        return remaining == 0 ? Status.OK : Status.BOUNDS_ERROR;
    }

    private Status redistributeSurplus(List<ProRataAllocator.ProRataSlice> slices, int surplus, int excludeAccount) {
        if (surplus <= 0) return Status.OK;
        int remaining = surplus;
        for (ProRataAllocator.ProRataSlice s : slices) {
            if (s.accountId == excludeAccount || remaining <= 0) continue;
            s.qtyMilli += 1;
            findSlice(slices, excludeAccount).qtyMilli -= 1;
            remaining--;
        }
        return remaining == 0 ? Status.OK : Status.BOUNDS_ERROR;
    }

    private int sumSliceQty(List<ProRataAllocator.ProRataSlice> slices) {
        int total = 0;
        for (ProRataAllocator.ProRataSlice s : slices) total += s.qtyMilli;
        return total;
    }

    private List<ProRataAllocator.ProRataSlice> deepCopy(List<ProRataAllocator.ProRataSlice> slices) {
        List<ProRataAllocator.ProRataSlice> out = new ArrayList<>();
        for (ProRataAllocator.ProRataSlice s : slices) {
            ProRataAllocator.ProRataSlice c = new ProRataAllocator.ProRataSlice();
            c.accountId = s.accountId;
            c.recordId = s.recordId;
            c.qtyMilli = s.qtyMilli;
            c.remainderRank = s.remainderRank;
            out.add(c);
        }
        return out;
    }
// --- expanded helpers ---
    public int computeSlack(List<ProRataAllocator.ProRataSlice> slices, List<AllocationBound> bounds) {
        int slack = 0;
        for (ProRataAllocator.ProRataSlice slice : slices) {
            AllocationBound b = findBound(bounds, slice.accountId);
            if (b != null) slack += Math.max(0, b.maxQtyMilli - slice.qtyMilli);
        }
        return slack;
    }

    public boolean isFeasible(List<AllocationBound> bounds, int targetTotal) {
        int minSum = 0, maxSum = 0;
        for (AllocationBound b : bounds) {
            minSum += b.minQtyMilli;
            maxSum += b.maxQtyMilli;
        }
        return targetTotal >= minSum && targetTotal <= maxSum;
    }
}

File: legacy-cpp/src/desk/allocation_solver.cc
#include "tkr/desk/allocation_solver.h"

#include <algorithm>
#include <unordered_map>
#include <vector>

namespace tkr {
namespace desk {
namespace {

std::uint32_t SumSliceQty(const std::vector<ProRataSlice>& slices) {
  std::uint32_t total = 0;
  for (const ProRataSlice& s : slices) {
    total += s.qty_milli;
  }
  return total;
}

const AllocationBound* FindBound(const std::vector<AllocationBound>& bounds,
                                 std::uint32_t account_id) {
  for (const AllocationBound& b : bounds) {
    if (b.account_id == account_id) {
      return &b;
    }
  }
  return nullptr;
}

struct BoundViolation {
  std::uint32_t account_id;
  std::uint32_t current_qty;
  std::uint32_t min_qty;
  std::uint32_t max_qty;
  std::int32_t deficit;
  std::int32_t surplus;
};

std::vector<BoundViolation> DetectViolations(
    const std::vector<ProRataSlice>& slices,
    const std::vector<AllocationBound>& bounds) {
  std::vector<BoundViolation> violations;
  for (const ProRataSlice& slice : slices) {
    const AllocationBound* bound = FindBound(bounds, slice.account_id);
    if (bound == nullptr) {
      continue;
    }
    BoundViolation v{};
    v.account_id = slice.account_id;
    v.current_qty = slice.qty_milli;
    v.min_qty = bound->min_qty_milli;
    v.max_qty = bound->max_qty_milli;
    if (slice.qty_milli < bound->min_qty_milli) {
      v.deficit = static_cast<std::int32_t>(bound->min_qty_milli - slice.qty_milli);
    }
    if (slice.qty_milli > bound->max_qty_milli) {
      v.surplus = static_cast<std::int32_t>(slice.qty_milli - bound->max_qty_milli);
    }
    if (v.deficit != 0 || v.surplus != 0) {
      violations.push_back(v);
    }
  }
  return violations;
}

}  // namespace

AllocationSolver::AllocationSolver(AllocationSolverConfig config)
    : config_(config) {}

void AllocationSolver::SetBound(std::uint32_t account_id, std::uint32_t min_qty,
                                std::uint32_t max_qty) {
  for (AllocationBound& b : custom_bounds_) {
    if (b.account_id == account_id) {
      b.min_qty_milli = min_qty;
      b.max_qty_milli = max_qty;
      return;
    }
  }
  AllocationBound bound{};
  bound.account_id = account_id;
  bound.min_qty_milli = min_qty;
  bound.max_qty_milli = max_qty;
  custom_bounds_.push_back(bound);
}

void AllocationSolver::ClearBounds() { custom_bounds_.clear(); }

Status AllocationSolver::ClampSlice(ProRataSlice* slice,
                                    const AllocationBound& bound) const {
  if (slice == nullptr) {
    return Status::kBoundsError;
  }
  if (slice->qty_milli < bound.min_qty_milli) {
    slice->qty_milli = bound.min_qty_milli;
  }
  if (slice->qty_milli > bound.max_qty_milli) {
    slice->qty_milli = bound.max_qty_milli;
  }
  return Status::kOk;
}

Status AllocationSolver::RedistributeDeficit(
    std::vector<ProRataSlice>* slices, std::uint32_t deficit_milli,
    std::uint32_t exclude_account) const {
  if (slices == nullptr || deficit_milli == 0) {
    return Status::kOk;
  }

  std::vector<std::size_t> donors;
  for (std::size_t i = 0; i < slices->size(); ++i) {
    if ((*slices)[i].account_id == exclude_account) {
      continue;
    }
    if ((*slices)[i].qty_milli > 0) {
      donors.push_back(i);
    }
  }

  if (donors.empty()) {
    return Status::kBoundsError;
  }

  std::uint32_t remaining = deficit_milli;
  std::size_t donor_idx = 0;
  std::uint32_t guard = 0;
  while (remaining > 0 && !donors.empty() && guard < deficit_milli * 4) {
    ProRataSlice& donor = (*slices)[donors[donor_idx % donors.size()]];
    if (donor.qty_milli > 0) {
      donor.qty_milli -= 1;
      remaining -= 1;
    }
    donor_idx += 1;
    guard += 1;
  }

  for (ProRataSlice& slice : *slices) {
    if (slice.account_id == exclude_account) {
      slice.qty_milli += deficit_milli - remaining;
      break;
    }
  }

  return remaining == 0 ? Status::kOk : Status::kBoundsError;
}

Status AllocationSolver::RedistributeSurplus(
    std::vector<ProRataSlice>* slices, std::uint32_t surplus_milli,
    std::uint32_t exclude_account) const {
  if (slices == nullptr || surplus_milli == 0) {
    return Status::kOk;
  }

  for (ProRataSlice& slice : *slices) {
    if (slice.account_id == exclude_account) {
      if (slice.qty_milli < surplus_milli) {
        return Status::kBoundsError;
      }
      slice.qty_milli -= surplus_milli;
      break;
    }
  }

  std::uint32_t remaining = surplus_milli;
  std::size_t idx = 0;
  std::uint32_t guard = 0;
  while (remaining > 0 && idx < slices->size() && guard < surplus_milli * 4) {
    if ((*slices)[idx].account_id != exclude_account) {
      (*slices)[idx].qty_milli += 1;
      remaining -= 1;
    }
    idx += 1;
    guard += 1;
  }

  return remaining == 0 ? Status::kOk : Status::kBoundsError;
}

bool AllocationSolver::CheckConvergence(
    const std::vector<ProRataSlice>& slices,
    const std::vector<AllocationBound>& bounds) const {
  return DetectViolations(slices, bounds).empty();
}

std::vector<AllocationBound> AllocationSolver::BuildDefaultBounds(
    const BatchWireFrame& frame) const {
  if (!custom_bounds_.empty()) {
    return custom_bounds_;
  }

  std::vector<AllocationBound> bounds;
  bounds.reserve(frame.records.size());
  for (const WireBatchRecord& rec : frame.records) {
    AllocationBound b{};
    b.account_id = rec.account_id;
    b.min_qty_milli = 0;
    b.max_qty_milli = rec.qty_milli * 2;
    bounds.push_back(b);
  }
  return bounds;
}

AllocationSolverResult AllocationSolver::SolveWithBounds(
    const std::vector<ProRataSlice>& initial,
    const std::vector<AllocationBound>& bounds,
    std::uint32_t total_qty_milli) {
  AllocationSolverResult result{};
  result.final_slices = initial;
  result.status = Status::kOk;

  if (!config_.enforce_bounds) {
    result.converged = true;
    return result;
  }

  for (std::uint32_t iter = 0; iter < config_.max_iterations; ++iter) {
    result.iterations_used = iter + 1;

    const std::vector<BoundViolation> violations =
        DetectViolations(result.final_slices, bounds);

    if (violations.empty()) {
      result.converged = true;
      break;
    }

    bool changed = false;
    for (const BoundViolation& v : violations) {
      for (ProRataSlice& slice : result.final_slices) {
        if (slice.account_id != v.account_id) {
          continue;
        }
        const AllocationBound* bound = FindBound(bounds, v.account_id);
        if (bound == nullptr) {
          break;
        }
        const std::uint32_t before = slice.qty_milli;
        ClampSlice(&slice, *bound);
        if (slice.qty_milli != before) {
          changed = true;
          result.bound_violations_fixed += 1;
        }
        break;
      }
    }

    const std::uint32_t current_total = SumSliceQty(result.final_slices);
    if (current_total < total_qty_milli) {
      const std::uint32_t deficit = total_qty_milli - current_total;
      for (const BoundViolation& v : violations) {
        if (v.deficit > 0) {
          RedistributeDeficit(&result.final_slices,
                              static_cast<std::uint32_t>(v.deficit),
                              v.account_id);
          changed = true;
          break;
        }
      }
      if (!changed) {
        RedistributeDeficit(&result.final_slices, deficit, 0);
      }
    } else if (current_total > total_qty_milli) {
      const std::uint32_t surplus = current_total - total_qty_milli;
      for (const BoundViolation& v : violations) {
        if (v.surplus > 0) {
          RedistributeSurplus(&result.final_slices,
                              static_cast<std::uint32_t>(v.surplus),
                              v.account_id);
          changed = true;
          break;
        }
      }
      if (!changed) {
        RedistributeSurplus(&result.final_slices, surplus, 0);
      }
    }

    if (!changed) {
      break;
    }
  }

  const std::uint32_t final_total = SumSliceQty(result.final_slices);
  if (final_total != total_qty_milli) {
    result.converged = false;
  } else {
    result.converged = CheckConvergence(result.final_slices, bounds);
  }

  return result;
}

AllocationSolverResult AllocationSolver::Solve(
    const BatchWireFrame& frame, const std::vector<ProRataSlice>& initial) {
  std::uint32_t total_qty = 0;
  for (const WireBatchRecord& rec : frame.records) {
    total_qty += rec.qty_milli;
  }

  const std::vector<AllocationBound> bounds = BuildDefaultBounds(frame);
  return SolveWithBounds(initial, bounds, total_qty);
}

}  // namespace desk
}  // namespace tkr
